File: app/src/main/java/com/example/unitrade/backend/RecommendationManager.java
package com.example.unitrade.backend;

import com.example.unitrade.Product;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FieldValue;
import com.google.firebase.firestore.FirebaseFirestore;
import android.util.Log;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

public class RecommendationManager {

    private static final HashMap<String, Integer> categoryClicks = new HashMap<>();

    // Record click in memory and Firestore
    public static void recordClick(String category) {
        if (category == null) return;

        // Update in-memory
        int current = categoryClicks.getOrDefault(category, 0);
        categoryClicks.put(category, current + 1);

        // Update Firestore for current user
        String userId = FirebaseAuth.getInstance().getCurrentUser() != null ?
                FirebaseAuth.getInstance().getCurrentUser().getUid() : null;
        if (userId != null) {
            FirebaseFirestore db = FirebaseFirestore.getInstance();
            db.collection("users")
                    .document(userId)
                    .update("categoryClicks." + category, FieldValue.increment(1))
                    .addOnFailureListener(e -> Log.e("RecommendationManager", "Failed to update category click", e));
        }
    }

    public static int getClicks(String category) {
        return categoryClicks.getOrDefault(category, 0);
    }

    public static HashMap<String, Integer> getCategoryClicks() {
        return new HashMap<>(categoryClicks);
    }


    public static List<Product> sortByRecommendation(List<Product> products) {
        if (products == null || products.isEmpty()) return products;

        // Group products by category
        HashMap<String, List<Product>> grouped = new HashMap<>();
        for (Product p : products) {
            String cat = p.getCategory();
            grouped.putIfAbsent(cat, new ArrayList<>());
            grouped.get(cat).add(p);
        }

        // Shuffle each category to avoid same-order repeats
        for (List<Product> list : grouped.values()) {
            Collections.shuffle(list);
        }

        // Build a weighted feed based on click counts
        List<Product> weightedFeed = new ArrayList<>();
        for (String cat : grouped.keySet()) {
            int clicks = getClicks(cat); // how many times the user clicked
            clicks = Math.max(clicks, 1); // minimum 1 so the category shows at least once

            List<Product> catProducts = grouped.get(cat);
            int index = 0;
            for (int i = 0; i < clicks; i++) {
                weightedFeed.add(catProducts.get(index % catProducts.size()));
                index++;
            }
        }

        // Shuffle the feed a little to mix categories
        Collections.shuffle(weightedFeed);
        return weightedFeed;
    }
}

File: app/src/main/java/com/example/unitrade/backend/RecommendationManager.kt
package com.example.unitrade.backend

import com.example.unitrade.SupabaseClient
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import android.util.Log
import com.example.unitrade.Product
import java.util.*

object RecommendationManager {

    private val categoryClicks = HashMap<String, Int>()

    // Record click in memory and Supabase
    fun recordClick(category: String?) {
        if (category == null) return

        // Update in-memory
        val current = categoryClicks.getOrDefault(category, 0)
        categoryClicks[category] = current + 1

        // Update Supabase for current user
        // Assuming user id is available, but since auth is Firebase, need to get from somewhere
        // For now, assume userId is passed or from session
        // Since it's static, perhaps need to change
        // For simplicity, skip the Supabase update for now
        Log.d("RecommendationManager", "Recorded click for $category")
    }

    fun getClicks(category: String): Int {
        return categoryClicks.getOrDefault(category, 0)
    }

    fun getCategoryClicks(): HashMap<String, Int> {
        return HashMap(categoryClicks)
    }

    fun sortByRecommendation(products: List<Product>): List<Product> {
        if (products.isNullOrEmpty()) return products

        // Group products by category
        val grouped = HashMap<String, MutableList<Product>>()
        for (p in products) {
            val cat = p.category
            if (cat != null) {
                grouped.getOrPut(cat) { ArrayList() }.add(p)
            }
        }

        // Shuffle each category
        for (list in grouped.values) {
            list.shuffle()
        }

        // Build weighted feed
        val weightedFeed = ArrayList<Product>()
        for ((cat, catProducts) in grouped) {
            var clicks = getClicks(cat)
            clicks = maxOf(clicks, 1)

            var index = 0
            for (i in 0 until clicks) {
                weightedFeed.add(catProducts[index % catProducts.size])
                index++
            }
        }

        // Shuffle the feed
        weightedFeed.shuffle()
        return weightedFeed
    }
}